lib/parallel/runner_test.go

```diff
@@ -0,0 +1,266 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package parallel
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"os/exec"
+	"sort"
+	"strconv"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestBasic(t *testing.T) {
+	cmds := []*exec.Cmd{
+		newSimpleCmd(0, "1", 0),
+		newSimpleCmd(0, "2", 0),
+		newSimpleCmd(0, "3", 0),
+		newSimpleCmd(0, "4", 0),
+		newSimpleCmd(0, "5", 0),
+	}
+	testEnv := newTestEnv(5, cmds)
+	require.NoError(t, testEnv.run())
+
+	testEnv.eventHandler.StartedEventSuccess(t)
+	testEnv.eventHandler.FinishedEventSuccess(t)
+	testEnv.eventHandler.NumEventsForTypeSuccess(t, EventTypeCmdStarted, 5)
+	testEnv.eventHandler.NumEventsForTypeSuccess(t, EventTypeCmdFinished, 5)
+	require.Equal(t, []string{"1", "2", "3", "4", "5"}, testEnv.stdout.SortedLines(t))
+}
+
+func TestError(t *testing.T) {
+	cmds := []*exec.Cmd{
+		newSimpleCmd(0, "1", 0),
+		newSimpleCmd(0, "2", 1),
+		newSimpleCmd(0, "3", 0),
+		newSimpleCmd(0, "4", 0),
+		newSimpleCmd(0, "5", 0),
+	}
+	testEnv := newTestEnv(5, cmds)
+	require.Error(t, testEnv.run())
+
+	testEnv.eventHandler.StartedEventSuccess(t)
+	testEnv.eventHandler.FinishedEventError(t)
+	testEnv.eventHandler.NumEventsForTypeSuccess(t, EventTypeCmdStarted, 5)
+	testEnv.eventHandler.NumEventsForTypeSuccess(t, EventTypeCmdFinished, 4)
+	testEnv.eventHandler.NumEventsForTypeError(t, EventTypeCmdFinished, 1)
+	require.Equal(t, []string{"1", "2", "3", "4", "5"}, testEnv.stdout.SortedLines(t))
+}
+
+func newSimpleCmd(sleepSec int, echoString string, exitCode int) *exec.Cmd {
+	return exec.Command(
+		"./testdata/bin/simple.sh",
+		strconv.Itoa(sleepSec),
+		echoString,
+		strconv.Itoa(exitCode),
+	)
+}
+
+type testEnv struct {
+	maxConcurrentCmds int
+	cmds              []*exec.Cmd
+	runner            *runner
+	eventHandler      *testEventHandler
+	stdout            *testBuffer
+	stderr            *testBuffer
+}
+
+func newTestEnv(maxConcurrentCmds int, cmds []*exec.Cmd) *testEnv {
+	stdout := newConcurrentReadWriter()
+	stderr := newConcurrentReadWriter()
+	for _, cmd := range cmds {
+		cmd.Stdout = stdout
+		cmd.Stderr = stderr
+	}
+	eventHandler := newTestEventHandler()
+	return &testEnv{
```

Comment:
This makes it a compile time check that all fields are filled, I think you can just look at the struct definition to see what is what in this block